From 72bc48996ee291045567dd8a014c2654cedc6110 Mon Sep 17 00:00:00 2001
From: Dennis Langenkamp <dennis.langenkamp@alliedvision.com>
Date: Tue, 22 Nov 2022 11:02:30 +0100
Subject: [PATCH 18/23] isi capture: Format changes for alvium cameras

- Added additional bayer and mono formats
- Added correct use of format alignment
- Fixed mbus format code for various rgb and yuv formats
---
 drivers/staging/media/imx/imx8-isi-cap.c |  16 +-
 drivers/staging/media/imx/imx8-isi-fmt.c | 348 ++++++++++++++++-------
 2 files changed, 255 insertions(+), 109 deletions(-)

diff --git a/drivers/staging/media/imx/imx8-isi-cap.c b/drivers/staging/media/imx/imx8-isi-cap.c
index f8673999f684..e265021e0ce2 100644
--- a/drivers/staging/media/imx/imx8-isi-cap.c
+++ b/drivers/staging/media/imx/imx8-isi-cap.c
@@ -950,14 +950,14 @@ static int mxc_isi_cap_try_fmt_mplane(struct file *file, void *fh,
     else
     {
         /*
-	 * The bit width in CHNL_IMG_CFG[HEIGHT/WIDTH] is 13, so the maximum
-	 * theorical value for image width/height should be 8K, but due to ISI
-	 * line buffer size limitation, the maximum value is 4K
-	 *
-	 * For efficient data transmission, the minimum data width should be
-	 * 16(128/8)
-	 */
-        v4l_bound_align_image(&pix->width, 16, ISI_4K, 2,
+         * The bit width in CHNL_IMG_CFG[HEIGHT/WIDTH] is 13, so the maximum
+         * theorical value for image width/height should be 8K, but due to ISI
+         * line buffer size limitation, the maximum value is 4K
+         *
+         * For efficient data transmission, the minimum data width should be
+         * 16(128/8)
+         */
+        v4l_bound_align_image(&pix->width, 16, ISI_4K, fmt->align,
                               &pix->height, 16, ISI_4K, 1, 0);
 
         pix->num_planes = fmt->memplanes;
diff --git a/drivers/staging/media/imx/imx8-isi-fmt.c b/drivers/staging/media/imx/imx8-isi-fmt.c
index e7c4aec0f395..2b425c25bb58 100644
--- a/drivers/staging/media/imx/imx8-isi-fmt.c
+++ b/drivers/staging/media/imx/imx8-isi-fmt.c
@@ -7,107 +7,253 @@
 #include "imx8-isi-core.h"
 
 struct mxc_isi_fmt mxc_isi_out_formats[] = {
-	{
-		.name		= "RGB565",
-		.fourcc		= V4L2_PIX_FMT_RGB565,
-		.depth		= { 16 },
-		.color		= MXC_ISI_OUT_FMT_RGB565,
-		.memplanes	= 1,
-		.colplanes	= 1,
-		.align		= 3,
-		.mbus_code  = MEDIA_BUS_FMT_RGB565_1X16,
-	}, {
-		.name		= "RGB24",
-		.fourcc		= V4L2_PIX_FMT_RGB24,
-		.depth		= { 24 },
-		.color		= MXC_ISI_OUT_FMT_BGR32P,
-		.memplanes	= 1,
-		.colplanes	= 1,
-		.align		= 2,
-		.mbus_code  = MEDIA_BUS_FMT_RGB888_1X24,
-	}, {
-		.name		= "BGR24",
-		.fourcc		= V4L2_PIX_FMT_BGR24,
-		.depth		= { 24 },
-		.color		= MXC_ISI_OUT_FMT_RGB32P,
-		.memplanes	= 1,
-		.colplanes	= 1,
-		.align		= 2,
-		.mbus_code  = MEDIA_BUS_FMT_BGR888_1X24,
-	}, {
-		.name		= "YUYV-16",
-		.fourcc		= V4L2_PIX_FMT_YUYV,
-		.depth		= { 16 },
-		.color		= MXC_ISI_OUT_FMT_YUV422_1P8P,
-		.memplanes	= 1,
-		.colplanes	= 1,
-		.align		= 3,
-		.mbus_code	= MEDIA_BUS_FMT_YUYV8_1X16,
-	}, {
-		.name		= "YUV32 (X-Y-U-V)",
-		.fourcc		= V4L2_PIX_FMT_YUV32,
-		.depth		= { 32 },
-		.color		= MXC_ISI_OUT_FMT_YUV444_1P8,
-		.memplanes	= 1,
-		.colplanes	= 1,
-		.align		= 2,
-		.mbus_code	= MEDIA_BUS_FMT_AYUV8_1X32,
-	}, {
-		.name		= "NV12 (YUYV)",
-		.fourcc		= V4L2_PIX_FMT_NV12,
-		.depth		= { 8, 8 },
-		.color		= MXC_ISI_OUT_FMT_YUV420_2P8P,
-		.memplanes	= 1,
-		.colplanes	= 2,
-		.align		= 4,
-		.mbus_code	= MEDIA_BUS_FMT_YUYV8_1X16,
-	}, {
-		.name		= "NV12M (YUYV)",
-		.fourcc		= V4L2_PIX_FMT_NV12M,
-		.depth		= { 8, 8 },
-		.color		= MXC_ISI_OUT_FMT_YUV420_2P8P,
-		.memplanes	= 2,
-		.colplanes	= 2,
-		.align		= 4,
-		.mbus_code	= MEDIA_BUS_FMT_YUYV8_1X16,
-	}, {
-		.name		= "YUV444M (Y-U-V)",
-		.fourcc		= V4L2_PIX_FMT_YUV444M,
-		.depth		= { 8, 8, 8 },
-		.color		= MXC_ISI_OUT_FMT_YUV444_3P8P,
-		.memplanes	= 3,
-		.colplanes	= 3,
-		.align		= 4,
-		.mbus_code	= MEDIA_BUS_FMT_YUV8_1X24,
-	}, {
-		.name		= "xBGR32",
-		.fourcc		= V4L2_PIX_FMT_XBGR32,
-		.depth		= { 32 },
-		.color		= MXC_ISI_OUT_FMT_XRGB32,
-		.memplanes	= 1,
-		.colplanes	= 1,
-		.align		= 2,
-		.mbus_code	= MEDIA_BUS_FMT_RGB888_1X24,
-	}, {
-		.name		= "ABGR32",
-		.fourcc		= V4L2_PIX_FMT_ABGR32,
-		.depth		= { 32 },
-		.color		= MXC_ISI_OUT_FMT_ARGB32,
-		.memplanes	= 1,
-		.colplanes	= 1,
-		.align		= 2,
-		.mbus_code	= MEDIA_BUS_FMT_RGB888_1X24,
-	},
-    //AVT additional format definition
-    {
-            .name		= "GREY",
-            .fourcc		= V4L2_PIX_FMT_GREY,
-            .depth		= { 8 },
-            .color		= MXC_ISI_OUT_FMT_RAW8,
-            .memplanes	= 1,
-            .colplanes	= 1,
-            .mbus_code	= MEDIA_BUS_FMT_Y8_1X8,
-    }
+        {
+                .name		= "RGB565",
+                .fourcc		= V4L2_PIX_FMT_RGB565,
+                .depth		= { 16 },
+                .color		= MXC_ISI_OUT_FMT_RGB565,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 3,
+                .mbus_code  = MEDIA_BUS_FMT_RGB888_1X24,
+        }, {
+                .name		= "RGB24",
+                .fourcc		= V4L2_PIX_FMT_RGB24,
+                .depth		= { 24 },
+                .color		= MXC_ISI_OUT_FMT_BGR32P,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code  = MEDIA_BUS_FMT_RGB888_1X24,
+        }, {
+                .name		= "BGR24",
+                .fourcc		= V4L2_PIX_FMT_BGR24,
+                .depth		= { 24 },
+                .color		= MXC_ISI_OUT_FMT_RGB32P,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code  = MEDIA_BUS_FMT_RGB888_1X24,
+        }, {
+                .name		= "YUYV-16",
+                .fourcc		= V4L2_PIX_FMT_YUYV,
+                .depth		= { 16 },
+                .color		= MXC_ISI_OUT_FMT_YUV422_1P8P,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 3,
+                .mbus_code	= MEDIA_BUS_FMT_UYVY8_2X8,
+        }, {
+                .name		= "YUV32 (X-Y-U-V)",
+                .fourcc		= V4L2_PIX_FMT_YUV32,
+                .depth		= { 32 },
+                .color		= MXC_ISI_OUT_FMT_YUV444_1P8,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_AYUV8_1X32,
+        }, {
+                .name		= "NV12 (YUYV)",
+                .fourcc		= V4L2_PIX_FMT_NV12,
+                .depth		= { 8, 8 },
+                .color		= MXC_ISI_OUT_FMT_YUV420_2P8P,
+                .memplanes	= 1,
+                .colplanes	= 2,
+                .align		= 4,
+                .mbus_code	= MEDIA_BUS_FMT_UYVY8_2X8,
+        }, {
+                .name		= "NV12M (YUYV)",
+                .fourcc		= V4L2_PIX_FMT_NV12M,
+                .depth		= { 8, 8 },
+                .color		= MXC_ISI_OUT_FMT_YUV420_2P8P,
+                .memplanes	= 2,
+                .colplanes	= 2,
+                .align		= 4,
+                .mbus_code	= MEDIA_BUS_FMT_UYVY8_2X8,
+        }, {
+                .name		= "YUV444M (Y-U-V)",
+                .fourcc		= V4L2_PIX_FMT_YUV444M,
+                .depth		= { 8, 8, 8 },
+                .color		= MXC_ISI_OUT_FMT_YUV444_3P8P,
+                .memplanes	= 3,
+                .colplanes	= 3,
+		        .align		= 4,
+                .mbus_code	= MEDIA_BUS_FMT_YUV8_1X24,
+        }, {
+                .name		= "xBGR32",
+                .fourcc		= V4L2_PIX_FMT_XBGR32,
+                .depth		= { 32 },
+                .color		= MXC_ISI_OUT_FMT_XRGB32,
+                .memplanes	= 1,
+                .colplanes	= 1,
+		        .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_RGB888_1X24,
+        }, {
+                .name		= "ABGR32",
+                .fourcc		= V4L2_PIX_FMT_ABGR32,
+                .depth		= { 32 },
+                .color		= MXC_ISI_OUT_FMT_ARGB32,
+                .memplanes	= 1,
+                .colplanes	= 1,
+		        .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_RGB888_1X24,
+        },
+        //AVT additional format definition
+        //Mono formats
+        {
+                .name		= "MONO8",
+                .fourcc		= V4L2_PIX_FMT_GREY,
+                .depth		= { 8 },
+                .color		= MXC_ISI_OUT_FMT_RAW8,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 1,
+                .mbus_code	= MEDIA_BUS_FMT_Y8_1X8,
+        },
+        {
+                .name		= "MONO10",
+                .fourcc		= V4L2_PIX_FMT_Y10,
+                .depth		= { 16 },
+                .color		= MXC_ISI_OUT_FMT_RAW10,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 1,
+                .mbus_code	= MEDIA_BUS_FMT_Y10_1X10,
+        },
+        {
+                .name		= "MONO12",
+                .fourcc		= V4L2_PIX_FMT_Y12,
+                .depth		= { 16 },
+                .color		= MXC_ISI_OUT_FMT_RAW12,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 1,
+                .mbus_code	= MEDIA_BUS_FMT_Y12_1X12,
+        },
+        //Bayer 8-bit
+        {
+                .name		= "BGGR8",
+                .fourcc		= V4L2_PIX_FMT_SBGGR8,
+                .depth		= { 8 },
+                .color		= MXC_ISI_OUT_FMT_RAW8,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_SBGGR8_1X8,
+        },
+        {
+                .name		= "GBRG8",
+                .fourcc		= V4L2_PIX_FMT_SGBRG8,
+                .depth		= { 8 },
+                .color		= MXC_ISI_OUT_FMT_RAW8,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_SGBRG8_1X8,
+        },
+        {
+                .name		= "GRBG8",
+                .fourcc		= V4L2_PIX_FMT_SGRBG8,
+                .depth		= { 8 },
+                .color		= MXC_ISI_OUT_FMT_RAW8,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_SGRBG8_1X8,
+        },
+        {
+                .name		= "RGGB8",
+                .fourcc		= V4L2_PIX_FMT_SRGGB8,
+                .depth		= { 8 },
+                .color		= MXC_ISI_OUT_FMT_RAW8,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_SRGGB8_1X8,
+        },
+        //Bayer 10-bit
+        {
+                .name		= "BGGR10",
+                .fourcc		= V4L2_PIX_FMT_SBGGR10,
+                .depth		= { 16 },
+                .color		= MXC_ISI_OUT_FMT_RAW10,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_SBGGR10_1X10,
+        },
+        {
+                .name		= "GBRG10",
+                .fourcc		= V4L2_PIX_FMT_SGBRG10,
+                .depth		= { 16 },
+                .color		= MXC_ISI_OUT_FMT_RAW10,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_SGBRG10_1X10,
+        },
+        {
+                .name		= "GRBG10",
+                .fourcc		= V4L2_PIX_FMT_SGRBG10,
+                .depth		= { 16 },
+                .color		= MXC_ISI_OUT_FMT_RAW10,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_SGRBG10_1X10,
+        },
+        {
+                .name		= "RGGB10",
+                .fourcc		= V4L2_PIX_FMT_SRGGB10,
+                .depth		= { 16 },
+                .color		= MXC_ISI_OUT_FMT_RAW10,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_SRGGB10_1X10,
+        },
+        //Bayer 12-bit
+        {
+                .name		= "BGGR12",
+                .fourcc		= V4L2_PIX_FMT_SBGGR12,
+                .depth		= { 16 },
+                .color		= MXC_ISI_OUT_FMT_RAW12,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_SBGGR12_1X12,
+        },
+        {
+                .name		= "GBRG12",
+                .fourcc		= V4L2_PIX_FMT_SGBRG12,
+                .depth		= { 16 },
+                .color		= MXC_ISI_OUT_FMT_RAW12,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_SGBRG12_1X12,
+        },
+        {
+                .name		= "GRBG12",
+                .fourcc		= V4L2_PIX_FMT_SGRBG12,
+                .depth		= { 16 },
+                .color		= MXC_ISI_OUT_FMT_RAW12,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_SGRBG12_1X12,
+        },
+        {
+                .name		= "RGGB12",
+                .fourcc		= V4L2_PIX_FMT_SRGGB12,
+                .depth		= { 16 },
+                .color		= MXC_ISI_OUT_FMT_RAW12,
+                .memplanes	= 1,
+                .colplanes	= 1,
+                .align		= 2,
+                .mbus_code	= MEDIA_BUS_FMT_SRGGB12_1X12,
+        },
 };
 
+
 size_t mxc_isi_out_formats_size = ARRAY_SIZE(mxc_isi_out_formats);
-- 
2.17.1

